tools/stub_generator/stub_codegen.py

```diff
@@ -0,0 +1,115 @@
+# Copyright 2020 The TensorFlow Runtime Authors
+# Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Lint as: python3
+"""Generates dynamic loading stubs for functions in CUDA and HIP APIs."""
+
+from __future__ import absolute_import
+from __future__ import print_function
+
+import argparse
+import json
+import clang.cindex
+
+
+def main():
+  parser = argparse.ArgumentParser(
+      description='Generate dynamic loading stubs for CUDA and HIP APIs.')
+  parser.add_argument('--unique_prefix', default="", type=str,
+                help='Unique prefix for used in the stub')
+  parser.add_argument(
+      'input', nargs='?', type=argparse.FileType('r'))
+  parser.add_argument(
+      'output', nargs='?', type=argparse.FileType('w'))
+  parser.add_argument(
+      'header', nargs='?', type=str, default=None)
+  parser.add_argument(
+      'extra_args', nargs='*', type=str, default=None)
+  args = parser.parse_args()
+
+  config = json.load(args.input)
+
+  function_impl = """
+{0}
+{{
+  using FuncPtr = {return_type} (%s *)({2});
+  static auto func_ptr = reinterpret_cast<FuncPtr>(load_symbol_func("{1}"));
+  if (!func_ptr) return {not_found_error};
```

Comment:
Just a thought: maybe we should generate a "not found" function with appropriate prototype and place it here in case of failure to load the symbol - this would save us the `if` in every successful call.

Reply:
Done